strawberry/schema/subscribe.py

```diff
@@ -0,0 +1,154 @@
+    ExecutionResult as OriginalExecutionResult,
+)
+from graphql.execution import subscribe as original_subscribe
+
+from strawberry.types import ExecutionResult
+from strawberry.types.execution import ExecutionContext, ExecutionResultError
+from strawberry.utils.await_maybe import await_maybe
+
+from .execute import (
+    ProcessErrors,
+    _coerce_error,
+    _handle_execution_result,
+    _parse_and_validate_async,
+)
+
+if TYPE_CHECKING:
+    from typing_extensions import TypeAlias
+
+    from graphql.execution.middleware import MiddlewareManager
+    from graphql.type.schema import GraphQLSchema
+
+    from ..extensions.runner import SchemaExtensionsRunner
+
+SubscriptionResult: TypeAlias = Union[
+    ExecutionResultError, AsyncIterator[ExecutionResult]
+]
+
+OriginSubscriptionResult = Union[
+    OriginalExecutionResult,
+    AsyncGenerator[OriginalExecutionResult, None],
+]
+
+
+async def _subscribe(
+    schema: GraphQLSchema,
+    execution_context: ExecutionContext,
+    extensions_runner: SchemaExtensionsRunner,
+    process_errors: ProcessErrors,
+    middleware_manager: MiddlewareManager,
+) -> AsyncGenerator[Union[ExecutionResultError, ExecutionResult], None]:
+    async with extensions_runner.operation():
+        if initial_error := await _parse_and_validate_async(
+            context=execution_context,
+            extensions_runner=extensions_runner,
+        ):
+            initial_error.extensions = await extensions_runner.get_extensions_results(
+                execution_context
+            )
+            yield await _handle_execution_result(
+                execution_context, initial_error, extensions_runner, process_errors
+            )
+        try:
+            async with extensions_runner.executing():
+                assert execution_context.graphql_document is not None
+                # Might not be awaitable if i.e operation was not provided with the needed variables.
+                agen_or_result: OriginSubscriptionResult = await await_maybe(
+                    original_subscribe(
+                        schema,
+                        execution_context.graphql_document,
+                        root_value=execution_context.root_value,
+                        variable_values=execution_context.variables,
+                        operation_name=execution_context.operation_name,
+                        context_value=execution_context.context,
+                        middleware=middleware_manager,
+                    )
+                )
+
+            # Handle immediate errors.
+            if isinstance(agen_or_result, OriginalExecutionResult):
+                yield await _handle_execution_result(
+                    execution_context,
+                    ExecutionResultError(data=None, errors=agen_or_result.errors),
+                    extensions_runner,
+                    process_errors,
+                )
+            else:
+                aiterator = agen_or_result.__aiter__()
```

Comment:
I comment below about using `aclosing`, this is another good place to use it

Reply:
how so? AsyncIterator has no `.aclose`?